test/integration/job/job_test.go

```diff
@@ -1174,6 +1174,263 @@ func TestBackoffLimitPerIndex(t *testing.T) {
+				if err != nil {
+					t.Fatalf("Failed to get the new Job: %v", err)
+				}
+				if diff := cmp.Diff(oldStatus, jobObj.Status); diff != "" {
+					t.Fatalf("Unexpected status (-want/+got): %s", diff)
+				}
+			}
+			if test.wantJobByExternalControllerTotalMetric != nil {
+				validateCounterMetric(ctx, t, metrics.JobByExternalControllerTotal, *test.wantJobByExternalControllerTotalMetric)
+			}
+		})
+	}
+}
+
+func TestManagedByLabel_RecreatedJob(t *testing.T) {
+	podTemplateSpec := v1.PodTemplateSpec{
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:                     "main-container",
+					Image:                    "foo",
+					ImagePullPolicy:          v1.PullIfNotPresent,
+					TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+				},
+			},
+		},
+	}
+
+	resetMetrics()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobManagedByLabel, true)()
+
+	closeFn, restConfig, clientSet, ns := setup(t, "managed-by-recreate-job")
+	defer closeFn()
+	ctx, cancel := startJobControllerAndWaitForCaches(t, restConfig)
+	defer func() {
+		cancel()
+	}()
+	baseJob := batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "custom-job-test",
+			Namespace: ns.Name,
+			Labels:    make(map[string]string),
+		},
+		Spec: batchv1.JobSpec{
+			Completions: ptr.To[int32](1),
+			Parallelism: ptr.To[int32](1),
+			Template:    podTemplateSpec,
+		},
+	}
+	jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &baseJob)
+	if err != nil {
+		t.Fatalf("Error %q while creating the job %q", err, jobObj.Name)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active:      1,
+		Ready:       ptr.To[int32](0),
+		Terminating: ptr.To[int32](0),
+	})
+	if err, _ := setJobPodsPhase(ctx, clientSet, jobObj, v1.PodSucceeded, 1); err != nil {
+		t.Fatalf("Failed setting phase %s on Job Pod: %v", v1.PodFailed, err)
+	}
+
+	validateJobSucceeded(ctx, t, clientSet, jobObj)
+
+	err = clientSet.CoreV1().Pods(jobObj.Namespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Error: '%v' while deleting the pods", err)
+	}
+
+	jobClient := clientSet.BatchV1().Jobs(jobObj.Namespace)
+	err = jobClient.Delete(ctx, jobObj.Name, metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatalf("Failed to delete job: %v", err)
+	}
+	jobObj, err = jobClient.Get(ctx, jobObj.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get deleted job: %v", err)
+	}
+
+	jobObj, err = updateJob(ctx, jobClient, jobObj.Name, func(j *batchv1.Job) {
+		j.Finalizers = []string{}
+	})
+	if err != nil {
+		t.Fatalf("Failed to clean up the orphan finalizer: %v", err)
+	}
+
+	// Wait for pods to start up.
+	err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, wait.ForeverTestTimeout, true, func(c context.Context) (done bool, err error) {
```

Comment:
This has been flagged as an antipattern in tests.

Reply:
What concretely? The use of active polling itself, or the short interval? We use active polling in many places, can you provide some reference?
I would be happy to use a watcher for this, but the code would be slightly more complex.

Reply:
hmm. I can't find the reference. maybe @pohly or @carlory know?

I know that for e2e tests I was swayed to not use PollUntilContextTimeout.

Reply:
Maybe I am mistaken. These were flagged in the e2e tests so not sure what we should do here.

Reply:
yes, the functions you mention are used in e2e tests where we have gomega framework. In the integration tests it is fairly common to use `PollUntilContextTimeout`, we have 24 occurrences in the `job_test.go`, for example: https://github.com/kubernetes/kubernetes/blob/0e8ef9c353092b4c921429a212ebb4c780f83715/test/integration/job/job_test.go#L2362